src/main/java/org/jsoup/select/NodeFilter.java

```diff
@@ -0,0 +1,82 @@
+package org.jsoup.select;
+
+import org.jsoup.nodes.Node;
+
+/**
+ * Node filter interface. Provide an implementing class to {@link NodeTraversor}
+ * to iterate through nodes.
+ * <p>
+ * This interface provides two methods, {@code head} and {@code tail}. The head
+ * method is called when the node is first seen, and the tail method when all of
+ * the node's children have been visited. As an example, head can be used to
+ * create a start tag for a node, and tail to create the end tag.
+ * </p>
+ * <p>
+ * For every node, the filter has to decide whether to
+ * <ul>
+ * <li>continue ({@link FilterResult#CONTINUE}),</li>
+ * <li>skip all children ({@link FilterResult#SKIP_CHILDREN}),</li>
+ * <li>skip node entirely ({@link FilterResult#SKIP_ENTIRELY}),</li>
+ * <li>remove the subtree ({@link FilterResult#REMOVE}),</li>
+ * <li>interrupt the iteration and return ({@link FilterResult#STOP}).</li>
+ * </ul>
+ * The difference between {@link FilterResult#SKIP_CHILDREN} and
+ * {@link FilterResult#SKIP_ENTIRELY} is that the first will invoke
+ * {@link NodeFilter#tail(Node, int)} on the node, while the latter does not.
+ * </p>
+ *
+ * @author Erich Schubert
+ */
+public interface NodeFilter {
+    /**
+     * Filter decision.
+     * 
+     * @author Erich Schubert
+     */
+    public enum FilterResult {
```

Comment:
I'd split this into two. One enum for the return value of `head()`, one for `tail()`. Being able to use values that semantically don't make sense feels like sloppy API design.